crates/ra_syntax/src/validation.rs

```diff
@@ -227,3 +228,16 @@ fn validate_visibility(vis: ast::Visibility, errors: &mut Vec<SyntaxError>) {
             .push(SyntaxError::new(SyntaxErrorKind::VisibilityNotAllowed, vis.syntax.text_range()))
     }
 }
+
+fn validate_range_expr(expr: ast::RangeExpr, errors: &mut Vec<SyntaxError>) {
+    let last_child = match expr.syntax().last_child_or_token() {
```

Comment:
I found it surprisingly hard to work with the syntax tree :( This happened to turn out okayish since I only care if the operator is at the end, but I don't know how I would in general find the range operator in a `RangeExpr`.

Reply:
That's because we don't have typed accesor yet for the range token. This should help

```diff
diff --git a/crates/ra_syntax/src/ast/expr_extensions.rs b/crates/ra_syntax/src/ast/expr_extensions.rs
index 25dbd0bed..b33d99b5a 100644
--- a/crates/ra_syntax/src/ast/expr_extensions.rs
+++ b/crates/ra_syntax/src/ast/expr_extensions.rs
@@ -189,6 +189,34 @@ impl ast::BinExpr {
     }
 }
 
+enum RangeOp {
+    /// `..`
+    Exclusive,
+    /// `..=`
+    Inclusive,
+}
+
+impl ast::RangeExpr {
+    fn op_details(&self) -> Option<(SyntaxToken, RangeOp)> {
+        self.syntax().children_with_tokens().filter_map(|it| it.into_token()).find_map(|c| {
+            let bin_op = match c.kind() {
+                T![..] => RangeOp::Exclusive,
+                T![..=] => RangeOp::Inclusive,
+                _ => return None,
+            };
+            Some((c, bin_op))
+        })
+    }
+
+    pub fn op_kind(&self) -> Option<RangeOp> {
+        self.op_details().map(|t| t.1)
+    }
+
+    pub fn op_token(&self) -> Option<SyntaxToken> {
+        self.op_details().map(|t| t.0)
+    }
+}
+
 impl ast::IndexExpr {
     pub fn base(&self) -> Option<ast::Expr> {
         children(self).nth(0)
```

Reply:
Oh, and 

```rust
    pub fn lhs(&self) -> Option<ast::Expr> {
        children(self).nth(0)
    }

    pub fn rhs(&self) -> Option<ast::Expr> {
        children(self).nth(1)
    }

    pub fn sub_exprs(&self) -> (Option<ast::Expr>, Option<ast::Expr>) {
        let mut children = children(self);
        let first = children.next();
        let second = children.next();
        (first, second)
    }
```

would be useful as well.

Reply:
Wait, no, there should be some additional logic to determine if a child is before or after the range token.

Basically, the AST API becomes bearable when all required typed accessors are added. We can't really use a more traditional enum approach, because everything is optional, and everything can contain whitespace.

Reply:
If you like, I can implement validation myself :)

Reply:
I tried to implement something here but it's not really great. Feel free to change it if you want :p